### ISSUE TITLE
Validation transform and async calls

### ISSUE THREAD

**Issue by ABIOLI-COM** [louthy/language-ext#876]
I checked back all the docs and all the issues, but I didn't find anything that can point me in the right direction.  
I wrote an 'inner' library, where the calls 'prepare' some data and return Validation<ValidationResult, TValue>.  
Let's call them f1, f2, f3, and so on...  
Now I use the library in a web service, where the result is a HttpResponseData.  
Any call can fail, and I want to be able to send back different status codes, so I wrote a few transformation functions, let's call them t1, t2, t3, and so on. This methods just convert the failure, leaving success untouched, so something like Validation<ValidationResult, TValue> to Validation<HttpResponseData, TValue>, where I match both cases and I transform the error side. 
Finally, some of the calls (let's say f2, for example) are async.  

An example of my situation (an Azure function) is something like:  

=>
    from v1 in f1().t2(req)  
    from v2 in f2().t3(req)  
    from v3 in f3().t1(req)
    select v3.TransformToResponse(req)  
    ...  

The TransformToResponse obviously matches both success and failures to HttpResponseData, so I end up with a Validation<HttpResponseData, HttpResponseData> result, that has to be matched (again) in both success and failure, because we have to return the exact value, based on the success or less of the whole operation.  

Moreover, if for example f2 is an async function, I cannot write this code:  

=>
    from v1 in f1().t2(req)  
    from v2 in (await f2()).t3(req)  
    from v3 in f3().t1(req)
    select v3.TransformToResponse(req)  
    ...  

The compiler complains with:  
error CS1995: The 'await' operator may only be used in a query expression within the first collection expression of the initial 'from' clause or within the collection expression of a 'join' clause  

I understand that this is a limitation of the compiler, but what we can do, for this case?  

Thanks in advance to everybody.
Andrea

**Comment by ABIOLI-COM** [comment]
I forgot the main question! :-) I feel that in general I could do better, than copy and paste, more or less, the same code... but I don't know how...

**Comment by louthy** [comment]
Unfortunately `Validation` is a synchronous type.  So trying to wrap asynchronous code with synchronous code is doomed to fail.  `ValidationAsync` hasn't been written yet, so there's nothing currently in the library that will help you.  

I note that you're using `Validation` with LINQ, and so it will early out on the first error (`Apply` is the function to use to collect all errors).  And so, this behaviour is the same as `Either`.  So you could switch to `EitherAsync`, which would support your use case above.

The best way to think of it is as a _'monadic onion'_.: the outer layers must have more capabilities than the inner layers.  So the inner most layer would be all pure functions, because they have no extra capabilities; then you step out to `Either`, which has pure-functions+an error state; then `EitherAsync` which has pure-functions+an error state+asynchronicity.

Finally, you may want to consider the new `Eff` and `Aff` monads.  Which are lazy synchronous and asynchronous operations that have an alternative value of `Error`.  The sync and async aspects of each play nice with each other.  They have lots more functionality for building IO heavy applications.   [I'm starting to build out the documentation for that](https://github.com/louthy/language-ext/wiki/How-to-deal-with-IO).

**Comment by ABIOLI-COM** [comment]
...your answers are always very 'juicy'... Thank you, I need to think a lot to what you are saying... :-)

**Comment by ABIOLI-COM** [comment]
I tried to follow the indications, substituting Validation with EitherAsync, but apparently I end up in the same hole.  

Let's suppose I have a 'AddCustomer' method that persists a new Customer, and that now returns a Task<EitherAsync<Seq<ValidationResult>, DbCustomer>>.  
In a test, I try to see if I can insert the same record twice (that's an error, of course).  
I wrote a test like:  
  
```
var q = from customer1 in await db.AddCustomer("Mario Rossi", "Some description", "Some address", true, true, "a@a.com", "Password1!")
                        from customer2 in await db.AddCustomer("Mario Rossi", "Some description", "Some address", true, true, "a2@a.com", "Password1!")
                        select customer2;

                await q.ItFailed(errors => Assert.Single(errors));
```  

But it complains with:  

`error CS1995: The 'await' operator may only be used in a query expression within the first collection expression of the initial 'from' clause or within the collection expression of a 'join' clause`

Did I do anything wrong? Can you have some other suggestion? Thanks again.

**Comment by ABIOLI-COM** [comment]
I have another question:

I have code that throws exceptions, when using **Validation<>**, handled correctly externally when called through **IfSuccess**.
After switching to **EitherAsync**, the corresponding (I think!) **IfRight** appears to 'swallow' the exception, that doesn't bubble anymore.
I tried to look in the source code, but I get lost in all the calls. Is it correct? Is it the expected behaviour?
Thanks again.

**Comment by louthy** [comment]
@ABIOLI-COM Could you post something more concrete?  There are a number of `IfRight` overloads.  Some that run actions, most that allow you to unify the right-bound-type to the left-bound-type.  It doesn't sound like that's what you want, but it would be good to understand a bit more context.

**Comment by ABIOLI-COM** [comment]
Ok, I'm sure I made a mess, forget about the IfRight swallowing exceptions, it should be my fault.
I remain with the doubt of the use of EitherAsync with async functions, when using it with LINQ to Objects (I cannot use an async after the first call): there are no samples/tests for using it, in the library.

**Comment by DanialK** [comment]
Hey @louthy 
Thanks for the great library.
I was wondering are there any plans to add `ValidationAsync` in the future versions by any chance?
Would be great to have a type that can accumulate async validation errors (like Validation).
I understand it is possible to implement this using EitherAsync and handle the accumulation ourselves but it would be quite nice to have a first class support with it, especially for use-cases where we'd want to combine our Sync and Async validations in our validation pipeline.
Thank you

**Comment by louthy** [comment]
@DanialK Eventually, yes.  I've been burning myself to the ground getting the `Aff`, `Eff`, and `Pipes` features ready for `v4` release, so it won't be super quick.  I'll need a bit of a break after this round of updates.  Feel free to have a go and submit a PR if you like.  You can take the cues from `EitherAsync` as to how it would work internally.

**Comment by louthy** [comment]
For those needing `ValidationAsync` like behaviour, take a look at [the latest applicative updates for `Eff` and `Aff`]((https://github.com/louthy/language-ext/releases/tag/v4.3.0-beta).  That will give the ability to collect multiple errors like `Validation`, whilst running expressions in parallel.

`ValidationAsync` will come eventually, just trying to pick my battles right now.

**Comment by louthy** [comment]
`ValidationT<F, IO, A>` is available [in version 5](https://github.com/louthy/language-ext/tree/main/LanguageExt.Core/Monads/Alternative%20Monads/ValidationT), which is the equivalent of the requested `ValidationAsync`.
